packages/editor/src/store/reducer.js

```diff
@@ -361,12 +432,80 @@ export const editor = flow( [
 		byClientId( state = {}, action ) {
 			switch ( action.type ) {
 				case 'SETUP_EDITOR_STATE':
-					return getFlattenedBlocks( action.blocks );
+					return getFlattenedBlocksWithoutAttributes( action.blocks );
+
+				case 'RECEIVE_BLOCKS':
+					return {
+						...state,
+						...getFlattenedBlocksWithoutAttributes( action.blocks ),
+					};
+
+				case 'UPDATE_BLOCK':
+					// Ignore updates if block isn't known
+					if ( ! state[ action.clientId ] ) {
+						return state;
+					}
+
+					// Do nothing if only attributes change.
+					const changes = omit( action.updates, 'attributes' );
+					if ( keys( changes ).length === 0 ) {
+						return state;
+					}
+
+					return {
+						...state,
+						[ action.clientId ]: {
+							...state[ action.clientId ],
+							...omit( action.updates, 'attributes' ),
+						},
+					};
+
+				case 'INSERT_BLOCKS':
+					return {
+						...state,
+						...getFlattenedBlocksWithoutAttributes( action.blocks ),
+					};
+
+				case 'REPLACE_BLOCKS':
+					if ( ! action.blocks ) {
+						return state;
+					}
+
+					return {
+						...omit( state, action.clientIds ),
+						...getFlattenedBlocksWithoutAttributes( action.blocks ),
+					};
+
+				case 'REMOVE_BLOCKS':
+					return omit( state, action.clientIds );
+			}
+
+			return state;
+		},
+
+		attributesByClientId( state = {}, action ) {
```

Comment:
Bikeshedding: The naming of the properties of `blocks` already doesn't feel very consistent, and I'm worried this moves us further in that direction. If `order` is also keyed by client IDs, why not name it `orderByClientId` ? Or conversely, name this one just `attributes` ? What _are_ the contents `byClientId` anyways, particularly relevant to the proposed changes of pulling some values out. It makes me wonder if we'd be better served by granularity on all of the properties, rather than some ambiguous "byClientId" block-ish object shape, i.e. 

```graphql
{
	blocks: {
		isValid, # Object<string, boolean>
		name, # Object<string, string>
		attributes, # Object<string, Object>
		order, # Object<string, Array>
	}
}
```

Reply:
I think this is probably a good idea :) each property is mapped by ClientID, making it easier to memoize selectors based on the properties used. Maybe better as a follow-up to keep the size of this PR contained?

Reply:
> Maybe better as a follow-up to keep the size of this PR contained?

I raised because I think the proposed changes shine light on the fact that we might be better served by considering _all_ of these block values in the same way (hash keyed by client ID), which could lead us to some solution for generalizing the reducer implementation rather than copy/paste-ing large parts of it across as we're doing here. I'd worry we'd continue this pattern in some possible future refactor if set as a precedent.

Reply:
I'm happy to make the change either in this PR or in a follow-up one. Let me know the final decision (when and which of the proposed conventions to go for) and I'll make it happen :)

Reply:
Let's try to do it here then.

Reply:
`originalContent` seems to also be a common property for a block. Does that need to be extracted to another by-client-id list? Are there any others I should know about?

If we're bikeshedding, let's be thorough :)

Reply:
As a more general comment, following the proposed approach would essentially encode the various properties of a block as separate branches in the state tree. The benefit is performance and separation, as discussed, but this comes at the cost of more difficulty in adding more properties to `block`, I would say.

As someone who is completely new to the project, I really don't know how to evaluate whether that's potentially a problem or not, but I wanted to bring it up as it occurred to me.

Reply:
> As a more general comment, following the proposed approach would essentially encode the various properties of a block as separate branches in the state tree. The benefit is performance and separation, as discussed, but this comes at the cost of more difficulty in adding more properties to `block` in the future (or changing them), I would say.

In my mind, I'd hoped it'd be some way we could programmatically generate the objects (reducers) based on some set of expected keys (`isValid`, `name`, `attributes`, and I suppose `originalContent` and any others), and that by doing it generically we would avoid any incurred cost in adding additional properties in the future, since it'd just be a key to add to the set.

This would assume that the logic for handling each key is consistent, which I'm not entirely confident is something we can assume (at least for attributes?). In which case, the refactoring may be more difficult and thus potentially not worthwhile. 

The big issue for me was in evaluating this part of state, how would you or anyone else answer the prompt "What is represented in the value of `byClientId`?" It's making less sense over time as we pull out separate properties to be able to call it as anything resembling a block object.

Reply:
More practically speaking, in my mind I'd had in mind some higher-order reducer akin to [`onSubKey`](https://github.com/WordPress/gutenberg/blob/80108fb9a2a5a2d3d5eb3692d0b47a65a9bb2073/packages/core-data/src/utils/on-sub-key.js), specific to handling the reducer actions relevant for blocks, on each block key.

So something roughly like:

```js
BLOCK_KEYS = [ 'name', 'isValid', 'originalContent', 'attributes' /* ? */ ]
blocks: combineReducers( BLOCK_KEYS.map( ( key ) => [ key, onBlockSubKey( key ) ] ) ) )
```

Reply:
> If `order` is also keyed by client IDs, why not name it `orderByClientId` ? Or conversely, name this one just `attributes` ?

Even if we're deferring a greater refactor, I'm still inclined to align these two. Thoughts on renaming to just `attributes`?

Reply:
Sure, no problem! I have no strong opinions there. Added a commit with the change.